### PR TITLE
Gh1460

### PR DESCRIPTION
**Pull request by FrankMittelbach**
# Internal housekeeping

## Status of pull request

- Ready to merge

## Checklist of required changes before merge will be approved
- [x] Test file(s) added
- [x] Version and date string updated in changed source files
- [x] Relevant `\changes` entries in source included
- [x] Relevant `changes.txt` updated
- [x] Rollback provided (if necessary)?
- [x] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

When combined with tagging code there are cases where one gets low-level errors after the real missing item error. That could perhaps be improved there, but it is an error situation and it is often the case that after a severe error the following errors are pretty weird, so I don't think it is too bad. The case is
```
\begin{itemize}
\begin{center} foo \end{center}
...
```
That generates a missing item error (the problem) and then low-level errors when tagging is activated.

### CODE REVIEW

File: base/doc/ltnews40.tex [hunk 1]
```diff
@@ -229,6 +229,24 @@ \section{Handling paragraph continuation}
...
+The commands \cs{addvspace} and \cs{addpenalty} generated the famous
+error message \enquote{Something's wrong--perhaps a missing \cs{item}}
+when they were encountered outside vertical mode. Most of the time this
+error was bogus and if not, then it was generated several times rather
+than once.
+
+Once upon a time (in \LaTeX{}~2.09) it was necessary that these
+commands were used only in vertical mode, but with \LaTeX2e in 1994,
+we changed the internals but simply overlooked that this error became
+useless already then. In this release, i.e. 30 years too late we have
```
Comment (josephwright) [inline base/doc/ltnews40.tex:243]:
```suggestion
commands were used only in vertical mode, but with \LaTeXe{} in 1994,
we changed the internals but simply overlooked that this error which
became useless then. In this release, i.e.,~30 years too late, we have
```

**Review by josephwright — CHANGES_REQUESTED** [review]
Some minor updates to the English suggested - other than that looks good-to-go

File: base/doc/ltnews40.tex (hunk 1 above)
Comment (josephwright) [reply base/doc/ltnews40.tex:243]: Suggestion doesn't quite auto-replace the right lines, but you get the idea!

Comment (FrankMittelbach) [reply base/doc/ltnews40.tex:243]: took me a whilem but just changes to one line I guess. bit weird the display. wonder what would happen if I would press "commit suggestion"

Comment (josephwright) [reply base/doc/ltnews40.tex:243]: Idea is to change lines 241-243, but somehow I've only managed to get GitHub to link this to line 243 :(

File: base/doc/ltnews40.tex [hunk 2]
```diff
@@ -229,6 +229,24 @@ \section{Handling paragraph continuation}
...
+The commands \cs{addvspace} and \cs{addpenalty} generated the famous
+error message \enquote{Something's wrong--perhaps a missing \cs{item}}
+when they were encountered outside vertical mode. Most of the time this
+error was bogus and if not, then it was generated several times rather
+than once.
+
+Once upon a time (in \LaTeX{}~2.09) it was necessary that these
+commands were used only in vertical mode, but with \LaTeX2e in 1994,
+we changed the internals but simply overlooked that this error became
+useless then. In this release, i.e., 30 years too late, we have
```
Comment (josephwright) [inline base/doc/ltnews40.tex:243]:
```suggestion
commands were used only in vertical mode, but with \LaTeXe{} in 1994,
we changed the internals but simply overlooked that this error which
became useless then. In this release, i.e.,~30 years too late, we have
```

**Review by josephwright — APPROVED** [review]

File: base/doc/ltnews40.tex (hunk 1 above)
Comment (FrankMittelbach) [reply base/doc/ltnews40.tex]: then obviously I didn't get it. please make the suggested changes directly on the branch file.

Comment (josephwright) [reply base/doc/ltnews40.tex]: I have - next time I'll get it right first time ;)

**Review by davidcarlisle — CHANGES_REQUESTED** [review]
I'm still not sure about the error checking `\@parmoderr` is for foloats which can't float out of par boxes : not in **outer** par mode  here being in inner v  mode is fine,. It does not giv ean error in a parbox (which is good) but teh error messagie implies that it would b eneeded. Als math mode gives th eprimitive missing `4` error (that's perhaps OK.

```
\documentclass{article}

\begin{document}

% no error
1 aaa\addvspace{1pt} bbb


% not in outer par mode
2 \mbox{aaa\addvspace{1pt} bbb}

% no error
3 \parbox{4cm}{aaa\addvspace{1pt} bbb}

% ! Missing $ inserted.
4 $aaa\addvspace{1pt} bbb$

\end{document}
```

**Comment by FrankMittelbach** [comment]
> I'm still not sure about the error checking `\@parmoderr` is for foloats which can't float out of par boxes : not in **outer** par mode here being in inner v mode is fine,. It does not giv ean error in a parbox (which is good) but teh error messagie implies that it would b eneeded. Als math mode gives th eprimitive missing `4` error (that's perhaps OK.

right, it should probably be getting its own error message, saying "not in par mode" or some other text. As to the extra errors, I think they are understandable even though primitive errors. One could avoid them by only giving an error is in restrict hmode, but it makes the coding more complicated without much gain in my opinion. (One could also avoid them by doing nothing if not in vmode, but I think the  ERROR  else `\par`  approach is actually better.

**Comment by FrankMittelbach** [comment]
or in LaTeX speak "Not allowed in LR-mode"

**Review by davidcarlisle — APPROVED** [review]

File: base/doc/ltnews40.tex [hunk 3]
```diff
@@ -229,6 +229,24 @@ \section{Handling paragraph continuation}
...
+The commands \cs{addvspace} and \cs{addpenalty} generated the famous
+error message \enquote{Something's wrong--perhaps a missing \cs{item}}
+when they were encountered outside vertical mode. Most of the time this
+error was bogus and if not, then it was generated several times rather
+than once.
+
+Once upon a time (in \LaTeX{}~2.09) it was necessary that these
+commands were used only in vertical mode, but with \LaTeXe{} in 1994,
+we changed the internals but simply overlooked that this error which
+became useless then. In this release, i.e.,~30 years too late, we have
```
Comment (car222222) [inline base/doc/ltnews40.tex:243]: Better: . . . that this error message then had become useless.

File: base/doc/ltnews40.tex [hunk 4]
```diff
@@ -229,6 +229,24 @@ \section{Handling paragraph continuation}
...
+finally lifted the ban and from now on this error should now only show
```
Comment (car222222) [inline base/doc/ltnews40.tex:244]: Too many "now"s!!

File: base/lterror.dtx [hunk 5]
```diff
@@ -766,6 +766,15 @@ Type  H <return>  for immediate help.}%
 %    \end{macrocode}
 % \end{macro}
 %
+% \begin{macro}{\@LRmoderr}
+% A command is not allowed in restrict horizontal mode, i.e., in LR-mode in \LaTeX{} terminology.
```
Comment (car222222) [inline base/lterror.dtx:770]: restricted

File: base/ltspace.dtx [hunk 6]
```diff
@@ -880,25 +861,54 @@
...
+% \changes{v1.3s}{2024/09/10}{Drop unnecessary \cs{@noitemerr} and
+%    instead generate \cs{@LRmoderr} if we are in restricted hmode (gh/1460)}
 %    \begin{macrocode}
 %</2ekernel>
 %<*2ekernel|latexrelease>
+%<latexrelease>\IncludeInRelease{2024/11/01}%
+%<latexrelease>                 {\addvspace}{drop unnecessary no-item error}%
+\protected\def\addvspace#1{%
+%    \end{macrocode}
+%    If encounter in hmode we check if we are in an hbox and if so
```
Comment (car222222) [inline base/ltspace.dtx:873]: When this is encountered in hmode, we check whether we are . . .

File: base/ltspace.dtx [hunk 7]
```diff
@@ -928,73 +938,116 @@
...
-%<latexrelease>\IncludeInRelease{2015/01/01}%
-%<latexrelease>                 {\addpenalty}{\addpenalty}%
+%<latexrelease>\IncludeInRelease{2024/11/01}%
+%<latexrelease>                 {\addpenalty}{\addpenalty drop error}%
 %<*2ekernel|latexrelease>
 %    \end{macrocode}
+%    \begin{macrocode}
+\protected\def\addpenalty#1{%
+%    \end{macrocode}
+%    See documentation of \cs{addvspace} for documentation of the next
```
Comment (car222222) [inline base/ltspace.dtx:952]: See the description of . . .

File: base/ltspace.dtx [hunk 8]
```diff
@@ -928,73 +938,116 @@
...
+        \penalty#1\relax
       \else
-        \ifdim\lastskip=\z@
-          \penalty#1\relax
-        \else
-          \@tempskipb\lastskip
+        \@tempskipb\lastskip
 %    \end{macrocode}
 %    We have to make sure the final \cs{vskip} seen by \TeX\ is the
 %    correct one, namely \cs{@tempskipb}. However we may have to
```
Comment (car222222) [inline base/ltspace.dtx:973]: However,

File: base/ltspace.dtx [hunk 9]
```diff
@@ -928,73 +938,116 @@
...
 %    adjust for \cs{prevdepth} when placing the penalty but that
```
Comment (car222222) [inline base/ltspace.dtx:974]: . . . penalty; but that . . .

**Review by car222222 — COMMENTED** [review]
Some tiresome details, as usual! 

I seem to be unable to approve this, or to request changes: why is this?

**Comment by josephwright** [comment]
> I seem to be unable to approve this, or to request changes: why is this?

Looks like it was already merged: I'll edit in your changes directly on develop

**Comment by FrankMittelbach** [comment]
> Some tiresome details, as usual!
> 
> I seem to be unable to approve this, or to request changes: why is this?

because it is not expected that each an everybody to approves or comments (because often people don't) so after 3 days without further comments it got merged. I can still add your suggested changes but the PR is no longer active thus you can't approve or request changes.

**Comment by FrankMittelbach** [comment]
> Looks like it was already merged: I'll edit in your changes directly on develop

appreciated

**Comment by car222222** [comment]
OK.  That sounds very reasonable.  Apologies for the delay.

File: required/latex-lab/testfiles-block/blocks-item-02.luatex.tlg [hunk 10]
```diff
@@ -75,13 +75,35 @@ l. ...  \begin{center} X
 Try typing  <return>  to proceed.
 If that doesn't work, type  X <return>  to quit.
 ==> blockenv common ending on input line ...
+! LaTeX Error: Something's wrong--perhaps a missing \item.
```
Comment (muzimuzhi) [inline required/latex-lab/testfiles-block/blocks-item-02.luatex.tlg:78]: Is this new no-item error expected?

It seems now
https://github.com/latex3/latex2e/blob/70890a831c8cdef4da25002e1de543312809ba54/required/latex-lab/testfiles-block/blocks-item-02.lvt#L29-L35
raises three no-item errors.

Comment (FrankMittelbach) [reply required/latex-lab/testfiles-block/blocks-item-02.luatex.tlg:78]: yes, well, let's say I accepted that as a side-effect for now. It is an error situation and it doesn't work that well in combination with tagging. I guess as some point one could revisit the mechanism and see how to make it better without slowing down the normal processing, but for now I think it is good enough.

**Pull request merged**